static/js/src/advantage/subscribe/react/APICalls/PendingPurchase.jsx

```diff
@@ -0,0 +1,77 @@
+import { useStripe } from "@stripe/react-stripe-js";
+import { useState } from "react";
+import { useQuery } from "react-query";
+import { getPurchase, postInvoiceID } from "../../../contracts-api";
+
+const usePendingPurchase = () => {
+  const [pendingPurchaseID, setPendingPurchaseID] = useState();
+
+  const stripe = useStripe();
+
+  const { isLoading, isError, isSuccess, data, error } = useQuery(
+    "pendingPurchase",
+    async () => {
+      const res = await getPurchase(pendingPurchaseID);
+
+      if (!res.stripeInvoices) {
+        throw new Error("Missing invoice");
+      }
+
+      const {
+        pi_decline_code,
+        pi_status,
+        pi_secret,
+        id: stripeInvoiceId,
+      } = res.stripeInvoices[0];
+
+      //Requires 3DS check
+      if (
+        pi_decline_code === "authentication_required" ||
+        (pi_status === "requires_action" && pi_secret)
+      ) {
+        const threeDSResponse = await stripe.confirmCardPayment(pi_secret);
+        if (threeDSResponse.error) {
+          const error = Error(threeDSResponse.error.message);
+          error.dontRetry = true;
+          throw error;
+        }
+      }
+
+      //Card declined
+      if (pi_status === "requires_payment_method") {
+        const invoiceRes = await postInvoiceID(
+          "purchase",
+          pendingPurchaseID,
+          stripeInvoiceId
+        );
+
+        const error = Error(JSON.parse(invoiceRes.errors).decline_code);
```

Comment:
You're mixing constructors with your thrown errors. But I'd prefer to never throw unless you want to application to halt, failing gracefully is preferred.